File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:

void solve(vector<string> &ans,int index,string &output,string &digits,vector<string> &mapping)
{
    if(index >= digits.length())
    {
        ans.push_back(output);
        return;
    }

    int digit = digits[index] - '0';
    string val = mapping[digit];
    for(int i = 0; i<val.length();i++)
    {
        char ch = val[i];
        output.push_back(ch);
        solve(ans,index+1,output,digits,mapping);
        output.pop_back();
    }
}
    vector<string> letterCombinations(string digits) {
        vector<string> ans;
        string output = "";
        int index = 0;
        if(digits.length() == 0)
        {
            return ans;
        }

        vector<string> mapping(10);
        mapping[2] = "abc";
         mapping[3] = "def";
          mapping[4] = "ghi";
           mapping[5] = "jkl";
            mapping[6] = "mno";
             mapping[7] = "pqrs";
              mapping[8] = "tuv";
               mapping[9] = "wxyz";

               solve(ans,index,output,digits,mapping);

               return ans;
        
    }
};

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        
        
        dic = {2:'abc',3:'def',4:'ghi',5:'jkl',6:'mno',7:'pqrs',8:'tuv',9:'wxyz'}
        ans = []
        digit = digits

        if digit == "":
            return []

        for i in digit:
            ans.append(int(i))
    
        res = []
        for i in ans:
            res.append(dic[i])
    
        final = []
        
        if len(res) == 1:
            for i in res[0]:
                final.append(i)
            return final

        elif len(res) == 2:
            for i in res[0]:
                for j in res[1]:
                    final.append(i+j)
            return final

        elif len(res) == 3:
            for i in res[0]:
                for j in res[1]:
                    for k in res[2]:                        
                        final.append(i+j+k)
            return final

        elif len(res) == 4:
            for i in res[0]:
                for j in res[1]:
                    for k in res[2]:
                        for l in res[3]:                        
                            final.append(i+j+k+l)
            return final
        